pkg/dao/rpm_test.go

```diff
@@ -263,3 +356,109 @@ func (s *RpmSuite) TestRpmSearchError() {
 	tx.RollbackTo(txSP)
 }
+
+type TestInsertForRepositoryCase struct {
+	given    int
+	expected string
+}
+
+var testCases []TestInsertForRepositoryCase = []TestInsertForRepositoryCase{
+	{
+		given:    scenario0,
+		expected: "empty slice found",
+	},
+	{
+		given:    scenario3,
+		expected: "",
+	},
+	{
+		given:    scenarioUnderThreshold,
+		expected: "",
+	},
+	{
+		given:    scenarioThreshold,
+		expected: "",
+	},
+	{
+		given:    scenarioOverThreshold,
+		expected: "",
+	},
+}
+
+func (s *RpmSuite) genericInsertForRepository(testCase TestInsertForRepositoryCase) {
+	t := s.Suite.T()
+	tx := s.tx
+
+	pagedRpmInsertsLimit := 100
+	dao := GetRpmDao(tx, map[string]interface{}{
+		OptionPagedRpmInsertsLimit: pagedRpmInsertsLimit,
+	})
+
+	p := s.preparePagedRpmInsert(testCase.given, pagedRpmInsertsLimit)
+	records, err := dao.InsertForRepository(s.repo.Base.UUID, p)
+
+	if testCase.expected != "" {
+		assert.Error(t, err)
+		assert.Contains(t, err.Error(), testCase.expected)
+	} else {
+		assert.NoError(t, err)
+		assert.Equal(t, int64(len(p)), records)
+	}
+}
+
+func (s *RpmSuite) TestInsertForRepositoryScenario0() {
+	s.genericInsertForRepository(testCases[scenario0])
+}
+
+func (s *RpmSuite) TestInsertForRepositoryScenario3() {
+	s.genericInsertForRepository(testCases[scenario3])
+}
+
+func (s *RpmSuite) TestInsertForRepositoryScenarioUnderThreshold() {
+	s.genericInsertForRepository(testCases[scenarioUnderThreshold])
+}
+func (s *RpmSuite) TestInsertForRepositoryScenarioThreshold() {
+	s.genericInsertForRepository(testCases[scenarioThreshold])
+}
+func (s *RpmSuite) TestInsertForRepositoryScenarioOverThreshold() {
+	s.genericInsertForRepository(testCases[scenarioOverThreshold])
+}
+
+func (s *RpmSuite) TestInsertForRepositoryWithExistingChecksums() {
+	t := s.Suite.T()
+	tx := s.tx
+
+	pagedRpmInsertsLimit := 100
+
+	dao := GetRpmDao(tx, map[string]interface{}{
+		OptionPagedRpmInsertsLimit: pagedRpmInsertsLimit,
+	})
+	p := s.preparePagedRpmInsert(scenarioThreshold, pagedRpmInsertsLimit)
+	records, err := dao.InsertForRepository(s.repo.Base.UUID, p[0:(pagedRpmInsertsLimit>>1)])
+	assert.NoError(t, err)
+	assert.Equal(t, records, int64(len(p[0:(pagedRpmInsertsLimit>>1)])))
+
+	records, err = dao.InsertForRepository(s.repo.Base.UUID, p[(pagedRpmInsertsLimit>>1):])
+	assert.NoError(t, err)
+	assert.Equal(t, records, int64(len(p[(pagedRpmInsertsLimit>>1):])))
+
+	records, err = dao.InsertForRepository(s.repo.Base.UUID, p[0:(pagedRpmInsertsLimit>>1)])
+	assert.NoError(t, err)
+	assert.Equal(t, records, int64(0))
+}
+
+func (s *RpmSuite) TestInsertForRepositoryWithWrongRepoUUID() {
+	t := s.Suite.T()
+	tx := s.tx
+
+	pagedRpmInsertsLimit := 100
+
+	dao := GetRpmDao(tx, map[string]interface{}{
+		OptionPagedRpmInsertsLimit: pagedRpmInsertsLimit,
+	})
+	p := s.preparePagedRpmInsert(scenario3, pagedRpmInsertsLimit)
+	records, err := dao.InsertForRepository(uuid.NewString(), p)
+
+	assert.Error(t, err)
+	assert.Equal(t, records, int64(0))
+}
```

Comment:
It would also be good to add a test that:

1.  inserts some set of rpms to a repo:  [a,b,c]
2.  does the same with a new set of semi-overlapping rpms:   [b,c,d]

and verify that `a` is no longer in the repo, but `d` is in the repo

Reply:
good test! and I think the current behavior does not match this; so when `InsertForRepository` is called on this scenario, `[b,c]` is preserved and not inserted again, `a` is removed because it is not in `[b,c,d]` set and `d` is added as it is not part of the rpms yet; am I right?

Reply:
yes exactly!    i imagine testing that b & c are not inserted again, might be difficult.  You could check the uuid of the rpm i guess.  I don't know that you could easily check that the repositories_rpms entry wasn't deleted and re-inserted though?

Reply:
no, I am checking the record counts; for the given scenario I expect some records inserted that could differ from the final rpm associated to the repository; but the numbers should match (4 new records inserted, but 5 records associated from 5 yum package records).